### ISSUE TITLE
Integrate ServiceWorkers into replay system

### ISSUE THREAD

**Issue by machawk1** [oduwsdl/ipwb#88]
This ticket's primary title is to correctly resolve absolutely defined embedded URIs to prevent fundamental archival leakage.

https://github.com/ibnesayeed/constructive via @ibnesayeed

# Task List
- [X] Include [sample WARC containing a page with an embedded resource](https://github.com/oduwsdl/ipwb/blob/master/ipwb/samples/warcs/froggie.warc.gz) included using an absolute URL. Be sure resource is also captured.
- [x] Take note of state of ipwb replay, should "leak" onto live web for this resource.
- [x] Tie-in service worker to ipwb replay.
  - [x] Finish memento endpoint in #105 (needed for embedded resource resolution with "closest" datetime)
  - [x] Write JavaScript to install, initialize, (etc) Service worker per https://github.com/ibnesayeed/constructive
  - [x] Integrate the current replay HTML/JS implementation with the ServiceWorker JS
- [ ] Write test cases for validating that the service worker intercepts the requests and a representation from the live web is not included in the replay system.

Relates to #6.

**Comment by machawk1** [comment]
JavaScript Natural Language date resolution library: http://wanasit.github.io/pages/chrono/

**Comment by machawk1** [comment]
Added `https://github.com/oduwsdl/ipwb/blob/master/ipwb/samples/warcs/froggie.warc.gz`as a capture of a page with a single image embedded using an absolute URL, which demonstrates leakage.

**Comment by machawk1** [comment]
- [X] Take note of state of ipwb replay, should "leak" onto live web for this resource.
![screen shot 2017-03-13 at 6 12 58 pm](https://cloud.githubusercontent.com/assets/2514780/23877462/ccb7bc9c-0818-11e7-81e7-baa4b2333b48.png)

**Comment by machawk1** [comment]
An approach I am using is having the serviceWorker.js file be under `webui/`, which means I have to set the `Service-Worker-Allowed: /` header, which works on the main replay page -- the service worker registers according to the console. However, no function actions (install, activate, fetch) are invoked when I navigation to a URI-M from the main replay page.

A few questions, @ibnesayeed :
1. Does the header need to be propagated to all responses from the replay system, even when viewing mementos?
2. Should we just have the serviceWorker be accessible from the root of the domain to simplify requests?
3. Any idea why the other listeners wouldn't be invoked on a memento? Does the memento have to actively call the `serviceWorker.js` events?

**Comment by ibnesayeed** [comment]
1. As far as I know `Service-Worker-Allowed` header is only necessary for the script resource (the server worker file) when the scope is to be customized. This is only utilized when the service worker is installed. Once the installation is done and scope is registered properly in the browser, the role of this header is over. No other resources need to have those headers. In fact it does not even make any sense because the fetch events need to be trapped at request time (before any response is received) also the resources might be hosted on a different domain where setting a custom header wont make sense.
2. If registration is causing an issue with custom scope, then we can begin with hosting the file in the root, but perhaps it would be more important to find out why it is failing. You might want to leverage [the examples of setting up custom scope](https://w3c.github.io/ServiceWorker/#service-worker-script-response).
3. Once the script is installed, only fetch event should be triggered when other resources in the scope are requested. Install and register events will only be fired when the page that registers the service worker is loaded.

**Comment by machawk1** [comment]
Despite setting the header on the JavaScript file, the scope remained the same. 

After some work I was able to shift files around and get the serviceWorker to execute when viewing mementos.

One reason I was only seeing the SW register but no further actions was resolved by testing in Incognito, indicating that the other actions had originally initially already occurred and just the re-registration process was reporting to the console.

Initial setup in https://github.com/oduwsdl/ipwb/tree/issue-88

**Comment by ibnesayeed** [comment]
One possibility could be the conflict with the already installed SW. Testing SW during development is tricky such as cleaning up all installed SWs. There are some configurations in the developer console to make it easier that can be leveraged.

**Comment by machawk1** [comment]
I found a couple instances in the Chrome UI where I could supposedly destroy the ServiceWorker, supposedly without avail, but it would be nice to have a sure-fire to do so from within Chrome. Alternatively, some code could be added to do this programmatically but testing in a clean Incognito made the process much simpler.

**Comment by machawk1** [comment]
![screen shot 2017-03-14 at 12 03 13 am](https://cloud.githubusercontent.com/assets/2514780/23885473/aff7a50a-0849-11e7-97c0-dd1bc167439c.png)
A CORS violation from a static absolutely defined image archived image attempting to hit the live web -- a good thing.

**Comment by ibnesayeed** [comment]
That is expected.

**Comment by machawk1** [comment]
The current service worker in 4f0a0eaa10f489f00d0529892ce6462d75a90d64, once registered, interferes with the webui assets.

**Comment by machawk1** [comment]
https://github.com/oduwsdl/ipwb/commit/c3d068e0c5321cc9468b4a6e1af011067f0bcc58 in the issue-88 branch now has the memento endpoint and service worker integration. The absolutely defined image in the above screenshot is intercepted by the serviceWorker, the URI transformed, and the "closest" URI-M of the image returned.

I will need to test this with a fresh head, but this ticket should be good to close.

**Comment by machawk1** [comment]
Closing this issue with 2112c26e2d88d0de9d00395d2c7db74cd09be763 . Test case requirement moved to #142.
